app/models/user.rb

```diff
@@ -75,6 +78,33 @@ def remove_devise_validations
     end
   end
 
+  def self.fetch_leaderboard_stats(current_user)
+    essential_fields  = [:id, :name, :profile_picture_url, :country_code, :wallet_balance]
+
+    # Fetches top users with maximum wallet_balance. Also appends their respective rank.
+    top_users_json    = User.order('wallet_balance desc').limit(LEADERBOARD_LIMIT).as_json(only: essential_fields)
+    top_users_json.each_with_index do |user, idx|
+      user['wb_rank'] = idx + 1
+    end
+
+    # Checks if current user is in top users
+    # If yes, then we are done.
+    # Else, finds rank of current user and build its json
+    current_user_in_top_users = top_users_json.select{|user| user['id'] == current_user.id}
+    if current_user_in_top_users.present?
+      current_user_json = current_user_in_top_users
+    else
+      current_user_json = current_user.as_json(only: essential_fields)
+      current_user_rank = User.where('wallet_balance > ?', current_user.wallet_balance).count + 1
+      current_user_json['wb_rank'] = current_user_rank
+    end
+
+    return {
+        top_users: top_users_json,
+        user: current_user_json
+    }
```

Comment:
We want even if the current user is in the top ten list to populate is ranking on the current_user_json
```suggestion
  essential_fields  = [:id, :name, :profile_picture_url, :country_code, :wallet_balance]

  # Fetches top users with maximum wallet_balance.
  top_users = User.order('wallet_balance desc').limit(LEADERBOARD_LIMIT)

  # Checks if current user is in top users
  # If yes, then we are done.
  # Else, finds rank of current user and build its json
  current_user_json = current_user.as_json(only: essential_fields)

  current_user_json['wb_rank'] = if top_users.include?(current_user)
                                   top_users.index(current_user) + 1
                                 else
                                   User.where('wallet_balance > ?', current_user.wallet_balance).count + 1
                                 end

  top_users_json.each_with_index do |user, idx|
    user['wb_rank'] = idx + 1
  end

  {
    top_users: top_users.as_json(essential_fields),
    user: current_user_json
  }
```

Reply:
In the suggested code if the current user is already in the top user list, won't we be calling `as_json` twice for this user instance : one while performing `current_user.as_json` and the other while performing `top_users.as_json` ?
I think you are concerned that if the current user is already in the top still the variable `current_user_json` needs to be populated, but I think it is already happening in the existing code. Please check ?